asset-transfer-basic/application-gateway-hsm/README.md

```diff
@@ -0,0 +1,121 @@
+
+```
+Error: Cannot find module 'pkcs11js'
+```
+
+how to install the required C Compilers and Python will depend on your operating system and version.
+
+## Install SoftHSM
+
+In order to run the application in the absence of a real HSM, a software
+emulator of the PKCS#11 interface is required.
+For more information please refer to [SoftHSM](https://www.opendnssec.org/softhsm/).
+
+SoftHSM can either be installed using the package manager for your host system:
+
+* Ubuntu: `sudo apt install softhsm2`
+* macOS: `brew install softhsm`
+* Windows: **unsupported**
+
+Or compiled and installed from source:
+
+1. install openssl 1.0.0+ or botan 1.10.0+
+2. download the source code from <https://dist.opendnssec.org/source/softhsm-2.5.0.tar.gz>
+3. `tar -xvf softhsm-2.5.0.tar.gz`
+4. `cd softhsm-2.5.0`
+5. `./configure --disable-gost` (would require additional libraries, turn it off unless you need 'gost' algorithm support for the Russian market)
+6. `make`
+7. `sudo make install`
+
+## Initialize a token to store keys in SoftHSM
+
+If you have not initialized a token previously (or it has been deleted) then you will need to perform this one time operation
+
+```bash
+echo directories.tokendir = /tmp > $HOME/softhsm2.conf
+export SOFTHSM2_CONF=$HOME/softhsm2.conf
+softhsm2-util --init-token --slot 0 --label "ForFabric" --pin 98765432 --so-pin 1234
+```
+
+This will create a SoftHSM configuration file called `softhsm2.conf` and will be stored in the home directory. This is
+where the sample expects to find a SoftHSM configuration file
+
+The Security Officer PIN, specified with the `--so-pin` flag, can be used to re-initialize the token,
+and the user PIN (see below), specified with the `--pin` flag, is used by applications to access the token for
+generating and retrieving keys.
+
+## Install PKCS#11 enabled fabric-ca-client binary
+To be able to register and enroll identities using an HSM you need a PKCS#11 enabled version of `fabric-ca-client`
+To install this use the following command
+
+```bash
+go get -tags 'pkcs11' github.com/hyperledger/fabric-ca/cmd/fabric-ca-client
+```
+## Enroll the HSM User
+
+A user, `HSMUser`, who is HSM managed needs to be registered then enrolled for the sample
+
+```bash
+cd scripts
+./generate-hsm-user.sh HSMUser
+```
+
+This will register a user `HSMUser` with the CA in Org1 (if not already registered) and then enroll that user which will
+generate a certificate on the file system for use by the sample. The private key is stored in SoftHSM
+
+### Go SDK
+
+For HSM support you need to ensure you include the `pkcs11` build tag.
+
+```
+cd application-gateway-hsm/go
+go run -tags pkcs11 hsm-sample.go
+```
+
+### Node SDK
+
+```
+cd application-gateway-hsm/node
+npm install
+npm run build
```

Comment:
It would probably be nice to include the build as a `prepare` script as with the other Gateway samples, so it happens automatically at the end of the `npm install` and not require an additional command to be run.